### ISSUE TITLE
Ambiguity of timestamp units

### ISSUE THREAD

**Issue by Banyc** [rust-pcap/pcap#370]
Not sure if this is a bug or a feature request but according to the [specs](https://ietf-opsawg-wg.github.io/draft-ietf-opsawg-pcap/draft-ietf-opsawg-pcap.html) searching with keyword `0xA1B2C3D4`, the timestamp in the packet headers can either be the following units:
- magic number `0xA1B2C3D4`: in seconds and microseconds
- magic number `0xA1B23C4D`: in seconds and nanoseconds

But the crate didn't expose the magic number field for use. Secondly, the packet header is just a simple memory casting

```rust
&*(&*header as *const raw::pcap_pkthdr as *const PacketHeader),
``` 

The name is `pub tv_usec: suseconds_t,` for the ambiguous part. Anyone didn't read the specs could end up using it as a microsecond even if the magic number happens to be `0xA1B23C4D`.

Given the timestamp field is explicitly defined in the specs, probably we can create a new struct with the unambiguous field name instead of using `libc::timeval`.

**Comment by Banyc** [comment]
Also the endian of the packet header fields relies on the endian of the magic number so the endian needs to be exposed as well.

**Comment by stappersg** [comment]
On Mon, Oct 21, 2024 at 11:56:33PM -0700, IchHabeKeineNamen wrote:
> so the endian needs to be exposed as well.
> 

patches welcome



Groeten
Geert "I do have a name" Stappers
https://www.catb.org/~esr/faqs/hacker-howto.html#style
-- 
Silence is hard to parse

**Comment by Banyc** [comment]
You are right that the only viable option is just exposing endian in the entire original post.

`libpcap` has already identified the endian of the savefile in [`pcap_check_header`, `sf-pcap.c`](https://github.com/the-tcpdump-group/libpcap/blob/dfc4a0a967e13d47d83ab67ea22248321dabdb9d/sf-pcap.c#L218) and handled the endian in [`pcap_next_packet`, `sf-pcap.c`](https://github.com/the-tcpdump-group/libpcap/blob/dfc4a0a967e13d47d83ab67ea22248321dabdb9d/sf-pcap.c#L517). The only intended way to choose precision is through `Capture<Offline>::from_file_with_precision()` (wrapper of `pcap_open_offline_with_tstamp_precision`). Given we have already had the current info we need, we don't need to learn the endian at all.

If we really really want to get the the endian for some reason without opening the file again, probably need to sneak in `pcap_t` (`struct pcap`) to read the `swapped` field, but that could involve a ton of unsafes and not worth the effort.

**Comment by stappersg** [comment]
On Thu, Oct 24, 2024 at 07:20:43AM -0700, IchHabeKeineNamen wrote:
> Closed #370 as completed.

Thank you
